lib/active_merchant/billing/gateways/flex_charge.rb

```diff
@@ -0,0 +1,327 @@
+        post[:environment] = @options[:tokenization_key] || @options[:access_token]
+      end
+
+      def evaluate(action, money, credit_card, options)
+        handle_evaluation(action, money, credit_card, options) unless options[:tokenize]
+
+        MultiResponse.run do |r|
+          r.process { tokenize(credit_card, options) }
+          extra_options = { payment_token: r.params.dig(:transaction, :payment_method) }
+          r.process { handle_evaluation(action, money, credit_card, options.merge(extra_options)) }
+        end
+      end
+
+      def handle_evaluation(action, money, credit_card, options)
+        post = {}
+        address = options[:billing_address] || options[:address]
+        add_merchant_data(post, options)
+        add_base_data(post, options)
+        add_invoice(post, money, options)
+        add_mit_data(post, money, credit_card, options)
+        add_payment(post, credit_card, address, options)
+        add_address(post, credit_card, address)
+        add_customer_data(post, options)
+        add_transaction_type(post, action)
+
+        commit(:purchase, post)
+      end
+
+      def tokenize(credit_card, options)
+        post = {}
+        add_credit_card(post, credit_card, options)
+
+        commit(:tokenize, post, headers)
+      end
+
+      def address_names(address_name, payment_method)
+        names = split_names(address_name)
+        return names if names.any?(&:present?)
+
+        [
+          payment_method&.first_name,
+          payment_method&.last_name
+        ]
+      end
+
+      def phone_from(options)
+        options[:phone] || options.dig(:billing_address, :phone) || options.dig(:billing_address, :phone_number)
+      end
+
+      def fetch_access_token
+        params = { AppKey: @options[:app_key], AppSecret: @options[:app_secret] }
+        url = build_request_url(:authenticate)
+        response = parse(ssl_post(url, params.to_json, headers))
+        @options[:access_token] = response[:accessToken]
+        @options[:expires] = response[:expires]
+      rescue ResponseError => e
+        raise OAuthResponseError.new(e)
+      end
+
+      def ensure_access_token_credential(try_again = true)
+        return if @options[:access_token] && @options[:expires] && ((@options[:expires] || 0) > DateTime.now.strftime('%Q').to_i)
+
+        fetch_access_token
+      rescue ActiveMerchant::OAuthResponseError => e
+        raise e
+      rescue ResponseError => e
+        raise e unless retry?(try_again, e)
+
+        @options.delete(:access_token)
+        ensure_access_token_credential(false)
+      end
+
+      def retry?(try_again, error)
+        try_again && %w(401 404).include?(error.response.code) && @options[:access_token].present?
+      end
+
+      def generate_uuid
+        SecureRandom.uuid
+      end
+
+      def build_request_url(action, id = nil)
+        "#{test? ? test_url : live_url}#{ENDPOINTS_MAPPING[action] % id}"
+      end
+
+      def headers
```

Comment:
💬 Comment / Style:

Just a matter of taste but I found forcing that merge not that clear, maybe tap can do the work more explicitly:

```ruby
def headers
  { 'Content-Type' => 'application/json' }.tap do |headers|
    headers['Authorization'] = "Bearer #{@options[:access_token]}" if @options[:access_token]
  end
end
```

Reply:
Done!